File: UnionFind/1970. Last Day Where You Can Still Cross/solution.py
class Solution:
    def latestDayToCross(self, ROW: int, COL: int, cells: List[List[int]]) -> int:
        parent = list(range(ROW*COL))

        def getID(r, c):
            return r*COL+c
        def getPos(id): # return (row, col)
            return id//COL, id%COL

        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x, y):
            px, py = find(x), find(y)
            if px == py: return
            if px <= py:
                parent[py] = px
            else:
                parent[px] = py

        n = len(cells)
        flooded = set([tuple((r-1, c-1)) for r, c in cells])
        dirs = [[-1,0],[1,0],[0,-1],[0,1]]
        for r in range(ROW):
            for c in range(COL):
                if (r,c) in flooded: continue
                for dr, dc in dirs:
                    row, col = r+dr, c+dc
                    if row<0 or row >= ROW or col < 0 or col >= COL: continue
                    if (row, col) in flooded: continue
                    union(getID(r,c), getID(row, col))

        for i in range(n-1, -1, -1):
            r, c = cells[i] # 1-based
            r, c = r-1, c-1
            flooded.remove((r, c))

            for dr, dc in dirs:
                row, col = r+dr, c+dc
                if row<0 or row >= ROW or col < 0 or col >= COL: continue
                if (row, col) in flooded: continue
                union(getID(r,c), getID(row, col))

            for col in range(COL):
                p = find(getID(ROW-1, col))
                row, _ = getPos(p)
                if row == 0: return i # 1-based, (i+1)-th day is still flooded, i-th day is the day
        return 0

File: UnionFind/1970. Last Day Where You Can Still Cross/solution.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <numeric>
#include <map>

using namespace std;

class Solution {
public:
    struct DSU {
        vector<int> parent;
        DSU(int n) {
            parent.resize(n);
            iota(parent.begin(), parent.end(), 0);
        }
        int find(int x) {
            if (parent[x] != x) parent[x] = find(parent[x]);
            return parent[x];
        }
        void unite(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            if (rootX != rootY) parent[rootX] = rootY;
        }
    };

    int latestDayToCross(int m, int n, vector<vector<int>>& cells) {
        // Grid to track land (0) vs water (1). 
        // We start with all water because we are going backwards in time.
        vector<vector<int>> grid(m, vector<int>(n, 1));
        
        DSU dsu(m * n + 2);
        int TOP = m * n;
        int BOTTOM = m * n + 1;
        int dirs[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

        // Iterate from the last day backwards
        for (int i = cells.size() - 1; i >= 0; --i) {
            int r = cells[i][0] - 1; // Convert 1-based to 0-based
            int c = cells[i][1] - 1;
            
            grid[r][c] = 0; // Turn water to land
            int key = r * n + c;

            // Connect to virtual TOP/BOTTOM nodes
            if (r == 0) dsu.unite(key, TOP);
            if (r == m - 1) dsu.unite(key, BOTTOM);

            // Connect to existing land neighbors
            for (auto& d : dirs) {
                int row = r + d[0];
                int col = c + d[1];
                if (row >= 0 && row < m && col >= 0 && col < n && grid[row][col] == 0) {
                    dsu.unite(key, row * n + col);
                }
            }

            // Check if a path exists
            if (dsu.find(TOP) == dsu.find(BOTTOM)) {
                return i;
            }
        }

        return 0;
    }
};